### PR TITLE
Light debug webserver

### PR DESCRIPTION
**Pull request by qdel**
Made in only one PR the 3 modifications:
- switch in sample app to use source map for debug builds
- modification on webserver.py file for serving files from source
- Added a capacity to switch the build mode easily when running scripts: (Release if not set)
 `$> BUILD_TYPE=RelWithDebInfo ./build_cmake.sh`
 `$> $env:BUILD_TYPE='Debug' ; .\build_cmake.bat`

I could separate the 3 commit, but they are all linked together.
(draft: testing on windows)

### CODE REVIEW

File: run_webserver.bat [hunk 1]
```diff
@@ -4,6 +4,6 @@ cd build
 
 echo Opening web browser and starting web server...
 
-start http://localhost:80/helloworld.html
+start http://localhost:8080/helloworld.html
 
-..\WebServer.py 80
+..\WebServer.py 8080
```
Comment (forderud) [inline run_webserver.bat]: Would you mind moving the changing of port to a _separate_ PR, so that we avoid bundling it with the web-server behavior changes?

I can preview & merge the port change quickly, so that you can rebase this PR on top of the changes afterwards.

File: WebServer.py [hunk 2]
```diff
@@ -1,13 +1,29 @@
 from http import server
 import sys
 
+match = {
+    '/source/': '../sample/'
```
Comment (forderud) [inline WebServer.py:5]: Would it be possible for you to add a comment that explains where the "source" and "sample" folder names comes from?

If you're referring to the "sample" folder in this repo, then I would prefer to avoid hardcoding that assumption in the script, since it would make the script incompatible with other projects.

File: image/Dockerfile [hunk 3]
```diff
@@ -72,4 +72,4 @@ RUN mkdir -p /project/dependencies/include && mkdir -p /project/dependencies/lib
 WORKDIR /project/build
 
 # Default build command will explicitly target wasm
-CMD /usr/local/Qt-wasm/bin/qt-cmake -DCMAKE_BUILD_TYPE=Release -G Ninja /project/source && ninja
+CMD /usr/local/Qt-wasm/bin/qt-cmake -DCMAKE_BUILD_TYPE=${BUILD_TYPE:-Release} -G Ninja /project/source && ninja
```
Comment (forderud) [inline image/Dockerfile:75]: Isn't the release vs. debug build mode already given by the Docker image, in the sense that it either contain a debug- or release-build of Qt?

I would then prefer that the application compiled here is always build in a consistent debug/release mode.

Comment (qdel) [reply image/Dockerfile:75]: There is no direct link between Qt and the app about debug and release configuration. (or in C/C++ in general when your link your app with external libraries)

You can have debug Qt and build your app in release and opposite, this could be even specified at the .cpp file compiled. If we want to keep it simple, you could easily create a Release version from a Debug version: just compile in debug and strip the debug symbols out from the binary.

Short version: you can debug your app, have the precise stack trace of the bug in your app without having Qt in debug mode. The compiler will show where it has debug symbol: the source code of your app.

(Remember that when you use the Qt installer, you download only the platform you want but can compile your app with all modes. If you want to debug Qt itself, you need to get Qt debug symbols from the installer.)

File: WebServer.py (hunk 2 above)
Comment (qdel) [reply WebServer.py:5]: We could define an environment variable specifying it. But it is one of the reason I wanted to put it inside the container: to avoid entering specificity of the project, these values are already hardcoded in the container.

I will add a comment anway to explain how it is used. (wording also is not pretty)

File: image/Dockerfile (hunk 3 above)
Comment (forderud) [reply image/Dockerfile:75]: > There is no direct link between Qt and the app about debug and release configuration. (or in C/C++ in general when your link your app with external libraries)

I know that's the case on Linux. However, on Windows you need to link to a debug-build of Qt when building an application in debug-mode, and a release-build of Qt when building an app in release-mode. The reason for that is that the memory layout (ABI) of certain C++ types differ between "debug" and "release" builds.

I would therefore like to follow the same pattern here, so that we stay consistent with existing practices on Windows.

File: WebServer.py (hunk 2 above)
Comment (forderud) [reply WebServer.py:5]: I would like to be able to copy&paste WebServer.py as-is into other codebases with similar need for testing WASM builds in a lightweight web-server.

File: image/Dockerfile (hunk 3 above)
Comment (qdel) [reply image/Dockerfile:75]: This is specific to msvc compiler only.

Doing this totally disables the user to debug his application as he will need to create a debug-specific container.

Thus, previous discussions about debug mode are obsoletes.

Comment (forderud) [reply image/Dockerfile:75]: > Doing this totally disables the user to debug his application as he will need to create a debug-specific container.

Correct. That's why I'm already creating a separate "debug" container for debug builds. There's also a separate "debug" branch in this repo for testing debug builds against the debug container.
